File: api/routers/recipes.py
from fastapi import APIRouter, Query, HTTPException
from api.dependencies import get_recipe_engine
from api.schemas.recipes import (
    RecipeByDishResponse,
    RecommendRequest,
    RecommendResponse,
    SubstitutionResponse,
    RecipeWithScore,
)

router = APIRouter()


@router.get("/by-dish", response_model=RecipeByDishResponse)
def get_recipe_by_dish(dish: str = Query(..., description="Dish name to look up")):
    engine = get_recipe_engine()
    recipe = engine.get_recipe_by_dish(dish, threshold=0.6)
    return RecipeByDishResponse(recipe=recipe)


@router.post("/recommend", response_model=RecommendResponse)
def recommend_recipes(body: RecommendRequest):
    engine = get_recipe_engine()
    raw = engine.recommend_by_ingredients(body.available_ingredients, top_k=body.top_k)
    recs = [RecipeWithScore(**r) for r in raw]
    return RecommendResponse(recommendations=recs)


@router.get("/substitutions", response_model=SubstitutionResponse)
def get_substitutions(ingredient: str = Query(...)):
    engine = get_recipe_engine()
    result = engine.get_substitutions(ingredient)
    if result is None:
        return SubstitutionResponse(ingredient=ingredient, alternatives=[], notes="")
    return SubstitutionResponse(
        ingredient=result["ingredient"],
        alternatives=result.get("alternatives", []),
        notes=result.get("notes", ""),
    )

File: frontend/lib/api/recipes.ts
import { get, post } from "./client";
import type { RecipeWithScore, SubstitutionResponse } from "@/lib/types";

interface RecipeByDishResponse {
  recipe: import("@/lib/types").Recipe | null;
}

interface RecommendResponse {
  recommendations: RecipeWithScore[];
}

export function getRecipeByDish(dish: string): Promise<RecipeByDishResponse> {
  return get<RecipeByDishResponse>(`/recipes/by-dish?dish=${encodeURIComponent(dish)}`);
}

export function recommendByIngredients(
  available_ingredients: string[],
  top_k = 3,
): Promise<RecommendResponse> {
  return post<RecommendResponse>("/recipes/recommend", { available_ingredients, top_k });
}

export function getSubstitutions(ingredient: string): Promise<SubstitutionResponse> {
  return get<SubstitutionResponse>(
    `/recipes/substitutions?ingredient=${encodeURIComponent(ingredient)}`,
  );
}
